docs/ldml/tr35.md

```diff
@@ -2972,6 +3033,40 @@ The binary operators '&', '-', and the implicit union have equal precedence and
 
 **One caution:** the '&' and '-' operators operate between sets. That is, they must be immediately preceded and immediately followed by a set. For example, the pattern **[[:Lu:]-A]** is illegal, since it is interpreted as the set **[:Lu:]** followed by the incomplete range **-A**. To specify the set of upper case letters except for 'A', enclose the 'A' in brackets: **[[:Lu:]-[A]]**.
 
+##### <a name="Variables_in_UnicodeSets" href="#Variables_in_UnicodeSets">Variables in UnicodeSets</a>
+
+Support for variable identifiers (var) is optional. 
+They are used in certain contexts such as in [https://cldr-smoke.unicode.org/spec/main/ldml/tr35-general.html#Transforms](Transforms).
+
+UnicodeSets may contain variables (`$my_char`, `$the_set`, ...) in place of full UnicodeSets and strings/characters.
+
+This introduces ambiguities when parsing. For example, it is unclear whether **[$a-$b]** is a range of strings/characters, or a set difference, because that depends on the values of the variables.
+In particular in the case of **$a-$b**, it must be that **$a** and **$b** are of the same kind, i.e., both are strings/characters or both are UnicodeSets. A variable map with any other combination (e.g., `{ a => '0', b => [:L:] }`) is incompatible with the syntax **$a-$b**, and would return an error at parse-time.
+Thus, implementations that support variables may want to require the value of variables to be known at UnicodeSet parse time.
+Parsing a UnicodeSet with variables support requires referenced variables to exist, in other words, parsing a UnicodeSet referring to an unknown variable leads to a parse error.
+
+Variables are replaced by value, that is, **[a \$minus z]** with a variable map `{ minus => '-' }` is equivalent to **[-az]**, not **[a-z]** (i.e., cardinality of 3 instead of 26).
```

Comment:
I believe that this is not done by ICU in transforms, so this would be needed to add to the list of changes that we would want in ICU.

Reply:
You are right, I added a (probably incomplete in minor ways) section to the PR description: "Required changes in ICU to reflect the new syntax".

In this case, CLDR data actually works the same with the ICU variable behavior and this draft's variable behavior. There are edge cases that work differently in current ICU and this draft, e.g., the example from the draft: `[a $minus z]` is `a, z, -` in this draft, and `a, b, c, ..., z` in ICU.

Reply:
That's what I was thinking, that ICU just does a variable substitution before parsing (the simplest implementation)

Reply:
Since we define 'types' of variables inside variable maps, I think that approach is not a valid implementation anymore. If we really wanted to allow implementations to pass in arbitrary strings of syntax characters, we'd need to define a new 'type' "SyntaxStream" or something like that. 

I am strongly against that (I believe @eggrobin is as well), but I understand that forcing implementation work onto ICU needs to be justified well. Theoretically speaking, until ICU decides that they do want proper variable support in UnicodeSets, they could just state that their UnicodeSet implementation _does not support variables as defined in UTS35 UnicodeSets_. As long as CLDR data adheres to UTS35, and both implementations of variables result in the same interpretation, that might be an option?

Reply:
Let's mull this over a bit. Since it is optional, we could also say that there are two optional profiles for extending US for variables: string substitution or syntactic substitution. We can encourage the latter, but allow for the former.

(Note: there is a ticket for formalizing conformance clauses in LDML, but we don't want to wait on that for your ticket.)

Reply:
Sounds good to me! 

Can I assume this current draft will (at least roughly) be a supported variant in the future? This is relevant for my implementation work in ICU4X. Small changes like whitespace significance here and there are no problem of course, but I'd like to confirm that this draft's typed variable approach will most likely _not_ conflict with any spec changes in the near future, and is safe to implement.

Reply:
> we could also say that there are two optional profiles for extending US for variables: string substitution or syntactic substitution

This is a terrible idea. The whole point of putting them in the grammar (and of this proposal) is to avoid this kind of interaction between the syntax and the variables; conversely, you cannot have variables within the grammar if they may do arbitrary syntactic things.

As Niels points out, if someone feels like doing string substitution, they can always do so at a higher level; the UnicodeSet syntax should not standardize higher-level syntaxes. After all, an implementation can always do `UnicodeSet(std::format("[a{}z]", "-"))`, and this is none of LDML’s business.

If you really want a profile, that profile should be in the transform rules syntax (that is the higher level syntax here), where instead of using UnicodeSet variables, an implementation could do string substitution.
But there, UnicodeSet is just the tip of the iceberg; if we start defining every weird edge case that Niels and I found playing with ICU’s implementation of the transform rules syntax, we’ll be writing profiles for another year.

Reply:
> if we start defining every weird edge case that Niels and I found playing with ICU’s implementation of the transform rules syntax, we’ll be writing profiles for another year.

(Off-topic for UnicodeSet, but fun example: https://util.unicode.org/UnicodeJsps/transform.jsp?a=%24meow+%3D+%28meo%2B%29w%3B%0D%0A%0D%0A%24meow%2B+%3E+%24meow%3B&b=meoooowmeoow%3Bmeowmeow%2Fmeow.)

Reply:
Ok, makes sense. BTW, I found that ICU implements the variables via SymbolTable, so it is not just text replacement. If we wanted to make ICU conform, we could (but possibly breaking compatibility with some external cliant).

Reply:
> If we wanted to make ICU conform, we could (but possibly breaking compatibility with some external cliant).

Yes, from what I can tell, a minimal compliance update would "just" have to update the UnicodeSet parser to make use of the SymbolTable in a less opaque way.

> BTW, I found that ICU implements the variables via SymbolTable, so it is not just text replacement.

Almost, it's a mix. In my opinion, that's even worse.

If you're interested, here's my understanding of how it works currently (in the context of transforms, using classes and names from ICU4J, though after a quick glance over ICU4C there's not much difference):

1. The SymbolTable that the transform rule parser uses is reused, in other words, variables are mapped to strings that are encodings of literal characters or some `UnicodeMatcher`s (`UnicodeSet` is a `UnicodeMatcher`), also encoded to characters (in the PUA)
2. The UnicodeSet parser uses a `RuleCharacterIterator`, which handles variables, to parse a UnicodeSet pattern.
3. A RuleCharacterIterator handles variable references, and *opaquely passes the string value of that variable on, character by character, without doing any decoding*
   * See [`next` that sets `this.buf`](https://github.com/unicode-org/icu/blob/2e45e6ec0e84a1c01812015a254ea31b286316fb/icu4j/main/classes/core/src/com/ibm/icu/impl/RuleCharacterIterator.java#L154)
   * See [`_current()` that uses `this.buf`](https://github.com/unicode-org/icu/blob/2e45e6ec0e84a1c01812015a254ea31b286316fb/icu4j/main/classes/core/src/com/ibm/icu/impl/RuleCharacterIterator.java#L338)
4. Back to UnicodeSet parser, this just receives characters from the iterator. Whenever it receives one, it tries to decode the character into a `UnicodeMatcher` that is specifically a `UnicodeSet`.
   * See [here](https://github.com/unicode-org/icu/blob/2e45e6ec0e84a1c01812015a254ea31b286316fb/icu4j/main/classes/core/src/com/ibm/icu/text/UnicodeSet.java#L2699)
5. If the decoding fails, this must be a regular character, so it is treated as a literal syntax character (**there is no distinction in RuleCharacterIterator's API whether a passed character comes from a variable or from the source string**). If it is a UnicodeSet, the UnicodeSet case is handled.

The takeaway from this should be that while variables pointing to UnicodeSets do have a (this-draft-compliant) special case handling, variables pointing to actual strings do not. The input streams `"x␣␣-␣b"` and `"x␣<variable pointing to "␣-␣b">"` are **indistinguishable** from the POV of the UnicodeSet parser (␣ as space).

A few fun examples:
* [Probably unrealistic](https://util.unicode.org/UnicodeJsps/transform.jsp?a=%24sample_uset+%3D+%27This+is+a+sample+UnicodeSet+%5Ba-z0-9%5D%27+%3B%0D%0A%0D%0A%23+I+think+this+should+replace+either+the+sample+text+or+the+character+%27%7C%27+with+MATCHED%2C+right%3F%0D%0A%5B%7C+%24sample_uset%5D+%3E+MATCHED+%3B&b=The+following+should+be+matched%3A+This+is+a+sample+UnicodeSet+%5Ba-z0-9%5D%0D%0AThe+following+should+be+matched%3A+%7C%0D%0A%0D%0AThis+should+not+match%3A+45458976)
* [Quite realistic, in my opinion](https://util.unicode.org/UnicodeJsps/transform.jsp?a=%24gamma_suff+%3D+%27-gamma%27+%3B%0D%0A%24alpha_suff+%3D+%27-alpha%27+%3B%0D%0A%0D%0A%23+remove+all+version+suffixes%2C+right%3F%0D%0A%5B%24alpha_suff+%24gamma_suff%5D+%3E+%3B%0D%0A%0D%0A%23+explanation%3A+above+is+equivalent+to+%5B-alpha-gamma%5D%2C+so+%5B%5C-+a-g+h+l+m+p%5D%0D%0A%0D%0A%23+%28working+version%3A%29%0D%0A%23+%24alpha_suff+%3E+%3B%0D%0A%23+%24gamma_suff+%3E+%3B&b=best_application-gamma)